src/cmd/ksh93/bltins/test.c

```diff
@@ -552,13 +553,27 @@ int sh_access(const char *name, int mode) {
             setreuid(shp->gd->euserid, shp->gd->userid) == 0) {
             mode = access(name, mode);
             // Restore ids.
-            if (shp->gd->userid != shp->gd->euserid) setreuid(shp->gd->userid, shp->gd->euserid);
+            if (shp->gd->userid != shp->gd->euserid) {
+                while (((status = setreuid(shp->gd->userid, shp->gd->euserid)) < 0) && (errno == EAGAIN));
+                if (status < 0) {
+                    // Restoring real user id failed, exit.
+                    error(ERROR_system(1), "setreuid(%d, %d) failed", shp->gd->userid, shp->gd->euserid);
+                }
+            }
             if (shp->gd->groupid != shp->gd->egroupid) {
-                setregid(shp->gd->groupid, shp->gd->egroupid);
+                while (((status = setregid(shp->gd->groupid, shp->gd->egroupid)) < 0) && (errno == EAGAIN)); 
+                if (status < 0) {
+                    // Restoring real group id failed, exit.
+                    error(ERROR_system(1), "setregid(%d, %d) failed", shp->gd->groupid, shp->gd->egroupid);
+                }
             }
             return mode;
         } else if (shp->gd->groupid != shp->gd->egroupid) {
-            setregid(shp->gd->groupid, shp->gd->egroupid);
+            while (((status = setregid(shp->gd->groupid, shp->gd->egroupid)) < 0) && (errno == EAGAIN));
+            if (status < 0) {
+                error(ERROR_system(0), "setregid(%d, %d) failed", shp->gd->groupid, shp->gd->egroupid);
```

Comment:
Why is the failure of `setreuid()` above fatal and the failure of `setregid()` here not fatal? It seems to me the two cases are symmetrical and should both be fatal or non-fatal.

Reply:
In the above case it fails while restoring real user id for the process.

Reply:
And here it fails restoring the real group id. How is that any less serious?

Reply:
Fixed.